### ISSUE TITLE
skip-appstream-check is ignored

### ISSUE THREAD

**Issue by ilya-fedin** [flathub-infra/flatpak-builder-lint#203]
I'm getting appstream-failed-validation despite having skip-appstream-check being set.

PS: what's the point of having appstream check two times?

**Comment by bbhtt** [comment]
The linter expects the flathub.json to be present in `builddir/build/files` but flatpak-builder doesn't copy anything there.

https://github.com/flathub/flatpak-builder-lint/blob/655d3fbeb53f11dcd54472851aec70472fc19451/flatpak_builder_lint/builddir.py#L112

If you add something in the manifest to copy it to `/app/flathub.json` it should work.

```yaml
- name: flathub-json
  buildsystem: simple
  build-commands:
    - cp flathub.json /${FLATPAK_DEST}/flathub.json
  sources:
  - type: file
    path: flathub.json
```

**Comment by bbhtt** [comment]
The update here is that appstream and icon check is no longer run in the flathub builders as a separate step. All checks are now in the linter, which checks the manifest and the repo once build is done.

baseapps, extensions and console applications will be allowed to have "skip-appstream-check" in `flathub.json` and/or skip the other appstream checks in linter. Baseapps can already skip through every check.

The `flathub.json` will be installed like above for the builddir and repo checks for every console-application on Flathub.

Regular applications cannot skip through some of this without exceptions or having `flathub.json` installed like above.

**Comment by ilya-fedin** [comment]
You mean having the exception for having the key in flathub.json is not enough? My case is regular application

**Comment by bbhtt** [comment]
> You mean having the exception for having the key in flathub.json is not enough? My case is regular application

Right now, no change is necessary.

**Comment by bbhtt** [comment]
https://github.com/flathub-infra/flatpak-builder-lint/pull/249 removed flathub.json handling for appstream checks. 

It isn't deployed to flathub yet so the checks won't take effect right now.

I assume you're talking about Telegram, with that PR it is failing:

```sh
{
    "errors": [
        "flathub-json-skip-appstream-check",
        "appstream-failed-validation",
        "appid-ends-with-lowercase-desktop"
    ],
    "warnings": [
        "appstream-screenshot-missing-caption"
    ],
    "appstream": [
        "style-invalid         : Too many <p> tags for a good description [12/10]"
    ]
}
```

You have to 

a) remove installing flathub.json to `/app` 
b) apply for an exception for `appid-ends-with-lowercase-desktop` (this can be granted based on the fact that telegram predated the linter)
c) either fix or apply for an exception for `appstream-failed-validation` (Not sure why appstream-util is being unnecessarily strict about p tag here) https://docs.flathub.org/docs/for-app-authors/linter#exceptions

**Comment by bbhtt** [comment]
Actually (c) should not be reached at all imo and should be relaxed to 20 (validate-relax default). 

There is already a patch in `org.flatpak.Builder` to relax it to 15 https://github.com/flathub/org.flatpak.Builder/blob/master/patches/appstream-glib-0005-Change-min.-paragraph-length-to-5-and-max.-number-of.patch but looks like it affects releases only.

I can add a patch to relax it for descriptions too.

**Comment by ilya-fedin** [comment]
> b) apply for an exception for `appid-ends-with-lowercase-desktop` (this can be granted automatically, based on the fact that telegram predated the linter) https://docs.flathub.org/docs/for-app-authors/linter#exceptions

But it is already existing: https://github.com/flathub-infra/flatpak-builder-lint/blob/master/flatpak_builder_lint/staticfiles/exceptions.json#L781

> Actually (c) should not be reached at all imo and should be relaxed to 15/20 (validate-relax default).

As far as I remember, the main problem with the linter is that it doesn't allow links in changelogs while Telegram changelogs frequently have links.

**Comment by bbhtt** [comment]
> As far as I remember, the main problem with the linter is that it doesn't allow links in changelogs while Telegram changelogs frequently have links.

Ok yea, looks like appstream-util doesn't want links in releases. Flathub patches it to allow in descriptions but not in releases.

Then you have to apply for the `appstream-failed-validation` exception or modify the workflow to use links within `<url type="details">URL</url>` inside `release` tags.

**Comment by ilya-fedin** [comment]
> Then you have to apply for the `appstream-failed-validation` exception

Does `skip-appstream-check` in flathub,json still has any effect? If yes, why the exception should be done two times?

**Comment by bbhtt** [comment]
> Does `skip-appstream-check` in flathub,json still has any effect? If yes, why the exception should be done two times?

Nope

**Comment by hfiguiere** [comment]
> Then you have to apply for the `appstream-failed-validation` exception

no. you patch it. releases have the `url` tag for that.

**Comment by hfiguiere** [comment]
and then you submit the changes upstream.

**Comment by ilya-fedin** [comment]
> no. you patch it. releases have the `url` tag for that.

I'm afraid I can't do that...

**Comment by bbhtt** [comment]
Looks like the spec says no plaintext url in release descriptions https://www.freedesktop.org/software/appstream/docs/sect-Metadata-Releases.html#tag-release-description and appstreamcli also errors on this. (The wording should be stronger on that, I'll open an issue)

I don't think it'd be sane to patch that out part. I'd also suggest using `<url>` or `<issues>`

**Comment by ilya-fedin** [comment]
Just to be clear, I don't understand the scripts that generate the appdata from changelog.txt in tdesktop and if it couldn't be fixed without changing tdesktop's scripts and exception couldn't be given then I will just give up on the package. tdesktop lacks contributors for anything but the functionality demanded by the business (flatpak is not one of them) so I'm doubt anyone will fix tdesktop.

**Comment by hfiguiere** [comment]
> > no. you patch it. releases have the `url` tag for that.
> 
> I'm afraid I can't do that...

You can modify any file.

**Comment by bbhtt** [comment]
I don't have an issue with the exception for "appstream-failed-validation" in general, if it is justified, considering it is effectively the replacement for "skip-appstream-check".

I looked at changelog.txt, there are 5 very old instances of lines having links like "open https://username.t.me" which fits neither `url` nor `issues` and nothing I found could be fitted into those two tags. Seems it is rare that an entry has a link.

My suggestion would be to strip any http(s) parts from changelog lines/replace them with a placeholder text either in the script or while building through patch/sed.

**Comment by hfiguiere** [comment]
> I don't have an issue with the exception for "appstream-failed-validation" in general,

There are only a few were it should be allowed. Malformed release entries isn't one of them.

**Comment by ilya-fedin** [comment]
> My suggestion would be to strip any http(s) parts from changelog lines/replace them with a placeholder text either in the script or while building through patch/sed.

Wouldn't that look weird to the users?

**Comment by bbhtt** [comment]
You could just remove the https part and preserve the rest, no software store renders it as clickable links if it is in release description.

**Comment by ilya-fedin** [comment]
But why should `https://` be stripped if the stores can interpret the text the way they want anyway?

**Comment by ilya-fedin** [comment]
The
```
- A new format was supported for username links, in addition to "t.me/username." You can now open Telegram accounts, groups or channels using links like "username.t.me" or "https://username.t.me."
```
Would become
```
- A new format was supported for username links, in addition to "t.me/username." You can now open Telegram accounts, groups or channels using links like "username.t.me" or "username.t.me."
```
Which is still weird

**Comment by hfiguiere** [comment]
>  contains a **brief description** of what is new in the release.

(emphasis mine)

So:

```
A new format was supported for username links, in addition to "t.me/username."
```

**Comment by ilya-fedin** [comment]
> So:

no one will do special changelogs for flatpak in tdesktop

> Because that's how appstream checks for hyperlinks.

Maybe it should stop doing the check if it doesn't have any sense?

**Comment by bbhtt** [comment]
> Maybe it should stop doing the check if it doesn't have any sense?

That you have to ask in https://github.com/ximion/appstream/issues I don't know the reasons, I'm following the spec and behaviour.

**Comment by bbhtt** [comment]
The linter changes were deployed yesterday and two telegram build happened since then that were successful.

Unless there are any specific issues with the linter, I think this can be closed.
